keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,870 @@
+users may want to consume services spread across clusters, but today that
+requires non-trivial bespoke solutions.
+
+The Multi-Cluster Services API aims to fix these problems.
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+- Define a minimal API to support service discovery and consumption across clusters.
+  - Consume a service in another cluster.
+  - Consume a service deployed in multiple clusters as a single service.
+- When a service is consumed from another cluster its behavior should be
+  predictable and consistent with how it would be consumed within its own
+  cluster.
+- Create building blocks for multi-cluster tooling.
+- Support multiple implementations.
+- Leave room for future extension and new use cases.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Define specific implementation details beyond general API behavior.
+- Change behavior of single cluster services in any way.
+- Define what NetworkPolicy means for multi-cluster services.
+- Solve mechanics of multi-cluster service orchestration.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+#### Terminology
+
+- **supercluster** - A placeholder name for a group of clusters with a high
+  degree of mutual trust and shared ownership that share services amongst
+  themselves. Membership in a supercluster is symmetric and transitive. The set
+  of member clusters are mutually aware, and agree about their collective
+  association.
+- **mcsd-controller** - A controller that syncs services across clusters and
+  makes them available for multi-cluster service discovery (MCSD) and
+  connectivitiy. There may be multiple implementations, this doc describes
+  expected common behavior.
+
+We propose a new CRD called `ServiceExport`, used to specify which services
+should be exposed across all clusters in the supercluster. `ServiceExports` must
+be created in each cluster that the underlying `Service` resides in. Creation of
+a `ServiceExport` in a cluster will signify that the `Service` with the same
+name and namespace as the export should be visible to other clusters in the
+supercluster.
+
+Another CRD called `ImportedService` will be introduced to store information
+about the services exported from each cluster, e.g. topology. This is analogous
+to the traditional `Service` type in Kubernetes. Each cluster will have a
+coresponding `ImportedService` for each uniquely named `Service` that has been
+exported within the supercluster, referenced by namespaced name.
```

Comment:
Can we consider making the import and export API Kinds match grammatically?

I and others have mistyped `ExportedService` instead of `ServiceExport` multiple times.

We could make them match:
- a )  `ServiceExport` && `ServiceImport`
- b )  `ExportedService` && `ImportedService`

Reply:
I like your suggestion of a) above, especially the matching prefix.

The thought process behind the current names was that they aren't exactly equivalent components:

`ServiceExport` is an export that acts on a `Service`, but doesn't contain any service-like configuration itself.
`ImportedService` is a full replacement for `Service` that happens to contain/reference imported data.

I agree that these are proving to be confusing and we could use better names, made worse since export/import are opposites and feel like they should be used in similar ways in their "opposite" resources.

Alternatively, since the import is a self-sufficient service, maybe it shouldn't just be defined as an "import". The key differentiator to a consumer is that it's a multi-cluster service and needs to be accessed as such. From that viewpoint multi-cluster is more descriptive than imported - imported from where? What do you think about replacing `ImportedService` with something like `MulticlusterService`?

Reply:
+1 to a

Reply:
Renamed ImportedService -> ServiceImport based on this conversation. Thanks!